### DISCUSSION TITLE
Using semantic-release with GitHub Actions

### DISCUSSION THREAD

**Discussion by electriquo** [semantic-release/semantic-release#2011] [Q&A]
## Current behavior

[Using semantic-release with GitHub Actions](https://github.com/semantic-release/semantic-release/blob/master/docs/recipes/github-actions.md#node-project-configuration) describes using semantic-release with `push` event

```yaml
on:
  push:
```

Switching to `pull_request` event, makes semantic-release output the following when running `semantic-release --debug --dry-run --no-ci`

```shell
[semantic-release] › ℹ  This test run was triggered on the branch refs/pull/23/merge, while semantic-release is configured to only publish from master, test, therefore a new version won’t be published.
```

## Expected behavior

Ability to run semantic-release on `pull_request` event

## Environment

- **semantic-release** version: 17.4.4
- CI environment: Github
- Plugins used: Attached in the configuration below
- **semantic-release** configuration: 
   ```js
   module.exports = {
     plugins: [
       "@semantic-release/commit-analyzer",
       ["@semantic-release/release-notes-generator", {
         preset: "conventionalcommits"
       }],
       "@semantic-release/changelog",
       "@semantic-release/git",
       "@semantic-release/github"
     ],
     tagFormat: "${version}",
     branches: [
       { name: 'master' },
       { name: currentBranch , prerelease: true }
     ]
   }
   ```

   where `currentBranch` is dynamically evaluated and in this specific case, it is evaluated to `test`

---

# Update

During the action, env-ci was integrated separately to display the collected values, and this is the output

```js
{
  isCi: true,
  name: 'GitHub Actions',
  service: 'github',
  commit: '57e7818dfca3dde8fba15e51fe6dc273083c81f6',
  isPr: true,
  branch: 'master',
  prBranch: 'refs/pull/23/merge',
  slug: 'foo/bar',
  root: '/home/runner/work/bar/bar',
  pr: 23
};
```

**Comment by travi** [comment]
could you explain your goals around 
* running against the PR build rather than a branch?
* using the `--dry-run` flag on a valid CI service?

semantic-release intentionally will not run on a PR build because most CI services (GitHub Actions included) run the PR build against a commit that merges the PR branch with the mainline branch. this is a commit that does not existing in any branch in the actual repository. 

normally this fails with an informative message, but using the `--dry-run` flag forces execution in this context that is not intended. since the commit does not exist outside of the PR build, no valid branch can be found.

**Reply by electriquo** [reply]
Firstly, thank you for the quick reply.
After diving into the code and debugging, you are 100% correct.

> could you explain your goals around
> * running against the PR build rather than a branch?
> * using the `--dry-run` flag on a valid CI service?

The goal is to leverage semantic-release in a continuous deployment methodology - for each branch that is intended to be merged (hence, trigger on `pull_request` event, rather than `push` event), create a prerelease. This will allow any dependent software to consume the prerelease and perform needed verification.

To achieve that, the prerelease branch configuration in `.rcrelease.js` is dynamically evaluated (as noted in the initial post).

`--dry-run` and `--debug` were added only for verification and they will be dropped once semantic-release will execute as intended.

**Reply by electriquo** [reply]
the issue described in the initial post was fixed.

The `env-ci` issue had the issue described in the initial post happened due to mismatching permissions. `env-ci` was not able to read the environment variables and event json file.

@travi: will update the initial post to reflect our discussion in https://github.com/semantic-release/semantic-release/discussions/2011#discussioncomment-968454

**Reply by electriquo** [reply]
> semantic-release intentionally will not run on a PR build because most CI services (GitHub Actions included) run the PR build against a commit that merges the PR branch with the mainline branch. this is a commit that does not existing in any branch in the actual repository.

@travi: While working on this matter, I had the same conclusion as you. to verify it, I wanted to compare the PR-branch and the branch where the PR originated from.

```shell
$ git config --add remote.origin.fetch +refs/pull/*/head:refs/remotes/origin/pr/*
$ git fetch --tags --all --prune
$ git diff origin/test origin/pr/23
$ echo $?
0
$ git ls-remote | grep 27f92cd5b3146946a7219234114aaeaddd6aef1c
From git@github.com:foo/bar.git
27f92cd5b3146946a7219234114aaeaddd6aef1c	refs/heads/test
27f92cd5b3146946a7219234114aaeaddd6aef1c	refs/pull/23/head
````

As you can see, there is no difference between the `test` branch and the pull-request, which seems to contradict what you said. Am I right that it contradicts or did I understand you incorrectly?

> since the commit does not exist outside of the PR build, no valid branch can be found.

What about `GITHUB_HEAD_REF` environment variable within the PR?

**Reply by travi** [reply]
> As you can see, there is no difference between the `test` branch and the pull-request, which seems to contradict what you said. Am I right that it contradicts or did I understand you incorrectly?

there is likely better documentation on the behavior somewhere, but i tracked [this](https://github.com/actions/checkout/blob/25a956c84d5dd820d28caab9f86b8d183aeeff3d/adrs/0153-checkout-v2.md#checkout-behavior) down.

> What about `GITHUB_HEAD_REF` environment variable within the PR?

that is github specific and would not enable flexibility of this _core_ behavior. this is common behavior across multiple CI services (for example, [Travis CI's behavior](https://docs.travis-ci.com/user/pull-requests/#how-pull-requests-are-built))

**Reply by travi** [reply]
> The goal is to leverage semantic-release in a continuous deployment methodology - for each branch that is intended to be merged (hence, trigger on `pull_request` event, rather than `push` event), create a prerelease. This will allow any dependent software to consume the prerelease and perform needed verification.

i would argue that you are over complicating your workflow. continuous deployment typically implies deploying stable changes in your project's mainline. this was one of the primary drivers for the initial creation of semantic-release.

semantic-release's pre-releases are intended for interating on experimental/unstable API changes that are not yet safe for general consumption. they are not intended to enable releasing untested behavior for testing within the consumer of the package. that is essentially hacking a gitflow-like workflow into semantic-release by misusing the pre-release functionality.

keep in mind that versions of your packages dont need to be tested in your consumers to be valid stable versions of your packages. if a version is found to be lacking a requirement once consumed, that becomes a new requirement of the package, potentially before the consumer actually moves forward with consuming it. implementation of that new requirement results in yet another stable version being published.

**Reply by electriquo** [reply]
thank you @travi for answering.

> that is github specific and would not enable flexibility of this _core_ behavior. this is common behavior across multiple CI services (for example, [Travis CI's behavior](https://docs.travis-ci.com/user/pull-requests/#how-pull-requests-are-built))

since [semantic-release utilizes env-ci](https://github.com/semantic-release/semantic-release/blob/143c8981bff89dfe92d6e6a2f0ffb8b7176c254d/index.js#L4), having specific [github thing](https://github.com/pvdlg/env-ci/blob/master/services/github.js), does not break anything.

> i would argue that you are over complicating your workflow. continuous deployment typically implies deploying stable changes in your project's mainline.

and yet, assume that all changes are stable, it does not mean that it does not need further testing before releasing it to the while. for instance, think about edge cases that cannot be easily auto tested, or one want to evaluate the performance impact due to the change, and there are many other scenarios.

to not steal more of your previous time, could you share how could semantic-release run on pull_request event and produce a prerelease?

**Reply by electriquo** [reply]
@travi: do you have any advice?

**Reply by travi** [reply]
> since [semantic-release utilizes env-ci](https://github.com/semantic-release/semantic-release/blob/143c8981bff89dfe92d6e6a2f0ffb8b7176c254d/index.js#L4), having specific [github thing](https://github.com/pvdlg/env-ci/blob/master/services/github.js), does not break anything.

i think i need to clarify my point from earlier a bit. 

> semantic-release intentionally will not run on a PR build because most CI services (GitHub Actions included) run the PR build against a commit that merges the PR branch with the mainline branch. this is a commit that does not existing in any branch in the actual repository.

this is core behavior that exists to account for the behavior that CI services commonly apply to their PR builds. adding an exception for one service is not something that we would entertain in core behavior, especially since the reason for the exception is to create an actual behavioral difference between services rather than just account for a variation in the context that needs to be handled. 

> The goal is to leverage semantic-release in a continuous deployment methodology - for each branch that is intended to be merged (hence, trigger on `pull_request` event, rather than `push` event), create a prerelease.

if you really want to enable pre-releases for all branches, you can match with `*`. i recognize the value of limiting this to those intended to be merged, but running semantic-release against a PR build just isnt a way we can enable to accomplish this. your best bet might be to configure the branch build to only run semantic-release if the branch has an associated PR open, but i leave figuring out how to accomplish that up to the reader. 

> and yet, assume that all changes are stable, it does not mean that it does not need further testing before releasing it to the while. for instance, think about edge cases that cannot be easily auto tested, or one want to evaluate the performance impact due to the change, and there are many other scenarios.

i agree that there are scenarios that may need additional verification against a published version, but these should be rare and limited to things that truly cannot be tested in isolation. for example, i use [storybook](https://storybook.js.org/) to isolate component libraries so that verification can be done against them rendered in a browser (and then i automate the verification using cypress). this way i can be very intentional about testing all edge cases w/o needing to publish and test within the consuming web-app.

i encourage you to find ways to do your common verification of the package behavior in isolation, outside of your consumer(s). it is important to consider your package as its own entity outside of any consumer. semantic-release is intended to help communicate the impact of the package changes. the more that is influenced by external forces, the less meaningful the semantic nature of the versions becomes.

**Reply by electriquo** [reply]
@travi: thank you for the clarification.

> if you really want to enable pre-releases for all branches, you can match with `*`

will try your offer and will update here.
